### DISCUSSION TITLE
TLS support

### DISCUSSION THREAD

**Discussion by samirkut** [Tochemey/goakt#499] [Q&A] [answered]
I am exploring the framework and it looks pretty nice. I had a question about the network stack. I want to use mutual TLS for all communication between different nodes. I cant seem to figure out how that would be injected in this case. Any pointers?

**Comment by Tochemey** [comment]
Hello @samirkut, I have not added any implementation or extension for TLS. I have been thinking about it, but I have not received any feature requests. Now, I think I have a valid use case for it.

**Reply by Tochemey** [reply]
I have created an issue #500 to tackle it. Hopefully to be done before the week ends :)

**Comment by Tochemey** [comment]
@samirkut I have started some work [here](https://github.com/Tochemey/goakt/pull/501). I hope to wrap it up before the week ends and QA it properly before releasing a new release.

**Comment by Tochemey** [comment]
@samirkut Even though I have started some work on the TLS support. I will need some time to read a bit around its performance. There are only two places where TLS may be needed:

- Remote calls: here it is easy to add TLS because we just need to know the CA of the remote node and we are done.
- Redistribution: where when a node dies its actors are recreated on the remaining nodes of the cluster as long as there is valid quorum to get going. In this scenario all nodes will need to know each other CA.

**Reply by samirkut** [reply]
So typically what I do is, I assume that all nodes share a CA. That is probably very likely and helps simplify the config quite a bit.

**Reply by Tochemey** [reply]
Sure we can go with that hypothesis. The other thing is to look into the handshake part for performance because we don't want to in any way hinder the current throughput (which need already some enhancement) when sending messages over the wire.

**Comment by Tochemey** [comment]
@samirkut I think I have a better to add the TLS. My only concern is the handshake part which will double the TCP connection time. That can be a performance hinderance. In any way I will document it once I wrap up the implementation.

**Comment by Tochemey** [comment]
@samirkut I have the PR ready #504. I will run some QA before merging it. However if you like to have it a go that will also be great. Thanks

**Reply by samirkut** [reply]
Awesome. Thanks so much. I plan to give it a go sometime today.

**Reply by Tochemey** [reply]
just use this commit hash [a505db1](https://github.com/Tochemey/goakt/pull/504/commits/a505db12925e9e33f79113f833c81cd23e0ed606)

**Reply by Tochemey** [reply]
@samirkut I have closed the branch because the tests did not produce the expected results and need more work to be done. I will revisit this feature in the Christmas break or another time. Thanks for bringing up though. Valid feature.

**Comment by Tochemey** [comment]
One a second thought I think you can establish secure connection with the following:

- Use a service mesh like istio or linkerd to establish secure communication in the cluster or amongst the various services or
- Put the cluster behind a VPC

Making using of TLS is great but I don't think it is necessary for the framework to support it looking at the various alternatives out there. However if there is anything that can help in the framework to support the alternatives, I am open to a suggestion.

**Comment by Tochemey** [comment]
@samirkut The elephant in the room with this implementation is that the underlying cluster engine library I am using does not support TLS from the first reading I did in the code base. So I will have to cater of that before we can really have a fully fledged mTLS in the whole system. This is the cluster engine I am using https://github.com/buraksezer/olric.


I can fork it and refactor it to the need of GoAkt to support mTLS. However this will take me some time and probably some sponsorship :)

**Comment by Tochemey** [comment, marked as answer]
At the moment this feature is not available and can be supported in the future.

**Comment by Tochemey** [comment]
@samirkut I have fully implemented the TLS support now. The same CA root is required for all nodes in the cluster for the TLS feature to work well. You can use the latest commit hash of https://github.com/Tochemey/goakt

**Reply by samirkut** [reply]
Awesome. I will give it a go hopefully this weekend
